Introducing the AdmissionPolicy blog (#64)
kubewarden/kubewarden.io: content/blog/2022/03/admission-policy.md

---
title: Introducing the AdmissionPolicy
authors:
- Raul Cabello Martin
date: 2022-03-09
---

> Up until now, the only way to define a policy in Kubewarden was to use a `ClusterAdmissionPolicy`. `ClusterAdmissionPolicies` are applied to matching cluster-wide resources and resources across all namespaces.

Reviewer: ```suggestion
Up until now, the only way to define a policy in Kubewarden was to use a `ClusterAdmissionPolicy` which is applied to cluster-wide resources across all namespaces.
```

> We are thrilled to announce `AdmissionPolicy`, a namespaced resource. The policy will process only the requests that are targeting the namespace where the `AdmissionPolicy` is defined. Other than that, there are no functional differences between `AdmissionPolicy` and `ClusterAdmissionPolicy`.

Reviewer: ```suggestion
That's why we're thrilled to announce the new `AdmissionPolicy` resource. This new resource is created inside a `namespace` and the policies will process only the requests that are targeting the namespace where the `AdmissionPolicy` is defined. Except from being a "namespaced" resource, AdmissionPolicy` works exactly the same as the `ClusterAdmissionPolicy`.
```

## Understanding the motivation for using AdmissionPolicies

> It is possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector`, but there is no way for Kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources just in a particular namespace. 

Reviewer: ```suggestion
It's possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector`, however there's no way for Kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources just in a particular namespace. 
```

Allowing all tenants to deploy `ClusterAdmissionPolicies` is risky. A tenant could apply policies that affect resources in all namespaces, even if they don't have access to all of them.

> You probably want to allow tenants to deploy policies just in the namespaces they have access to. That's where `AdmissionPolicies` are useful! We can achieve this restriction with `AdmissionPolicies` and Kubernetes’ Role-Based Access Control (RBAC).

Reviewer: ```suggestion
You probably want to allow tenants to deploy policies just in the namespaces they have access to. That's where `AdmissionPolicies` come into play! We can achieve this restriction with `AdmissionPolicies` and Kubernetes’ Role-Based Access Control (RBAC).
```

> Imagine we have a tenant who can deploy resources just to the `development` namespace. We can allow this tenant to deploy `AdmissionPolicies` just in this namespace using RBAC. If we allowed them to create `ClusterAdmissionPolicies`, they could block other resources from being handled in other namespaces.

Reviewer: ```suggestion
As an example, let's say you have a tenant who can deploy resources just to the `development` namespace. You can allow this tenant to deploy `AdmissionPolicies` just in this namespace using RBAC. If you allowed them to create `ClusterAdmissionPolicies`, they could block other resources from being handled in other namespaces.
```

## Let's see an AdmissionPolicy in Action!

We are going to create an `AdmissionPolicy` that rejects privileged pods from being created in the `production` namespace.

> Assume we have a Kubernetes cluster with Kubewarden installed. Follow the [quick start guide](https://docs.kubewarden.io/quick-start.html) if you don't have Kubewarden installed.

Reviewer: ```suggestion
For this example, a Kubernetes cluster with Kubewarden already installed is required. The installation process is described in the [quick start guide](https://docs.kubewarden.io/quick-start.html).
```

First, create two namespaces: `development` and `production`

```
kubectl create ns development 
kubectl create ns production
```

> Let's create an AdmissionPolicy in the `production` namespace:

Reviewer: ```suggestion
Once the namespaces are created, create a new `AdmissionPolicy` resource in the `production` namespace:
```

```yaml
apiVersion: policies.kubewarden.io/v1alpha2
kind: AdmissionPolicy
metadata:
  name: ns-privileged-pods
  namespace: production
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.1.9
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
  mutating: false

```

> Let's wait for the policy to be active:

Reviewer: ```suggestion
Wait for the policy to be active:
```

```
kubectl wait --for=condition=PolicyActive admissionpolicies ns-privileged-pods -n production
```

Create a file named `privileged-pod.yaml` with the following pod specification:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: privileged-pod
spec:
  containers:
    - name: nginx
      image: nginx:latest
      securityContext:
        privileged: true
```

> If we try to create a privileged pod in the `production` namespace:

Reviewer: ```suggestion
Try to create a privileged pod in the `production` namespace:
```

```
kubectl apply -f privileged-pod.yaml -n production
```

> we will get the following error:

Reviewer: ```suggestion
You will get the following error:
```

```
Error from server: error when creating "privileged-pod.yaml": admission webhook "namespaced-kubewarden-privileged-pods.kubewarden.admission" denied the request: User 'my-user' cannot schedule privileged containers
```

> Finally, verify we can successfully create a privileged pod in a different namespace:

Reviewer: ```suggestion
Finally, verify you can successfully create a privileged pod in a different namespace:
```

```
kubectl apply -f privileged-pod.yaml -n development
```

## Conclusion

> Do you want your policy to be applied to resources across all namespaces or cluster-wide resources? Then use `ClusterAdmissionPolicies`. 

Reviewer: ```suggestion
As you could see, Kubewarden can now provide you with two different choices for deploying policies.

If your policy needs to be applied to resources across all namespaces or cluster-wide resources, then you can use `ClusterAdmissionPolicies`. 
```

> Is your cluster shared by multiple users or teams? Are they using different namespaces? Do you want your policy to be applied to resources within a namespace? Then you can now take advantage of the new `AdmissionPolicy`.

Reviewer: ```suggestion
In the other hand, if your cluster is shared by multiple users or teams, using different namespaces or your policy needs to be applied only to resources within a namespace, then the new `AdmissionPolicy` will be the right choice.
```

> You can find the `AdmissionPolicy` specification [here](https://github.com/kubewarden/kubewarden-controller/blob/main/docs/crds/README.asciidoc#admissionpolicy).

Reviewer 2: If you change the link to https://github.com/kubewarden/kubewarden-controller/blob/e0433fc3774d06dcf5e08bf2c600ad0117b89448/docs/crds/README.asciidoc#admissionpolicy, it will be more future proof :)

Just in case we remove the crd docs generation in-tree in the future, or something similar.

Author: done https://github.com/kubewarden/kubewarden.io/pull/64/commits/987dacff485120ed208e01de98f752b0221e0813

> You can find the `AdmissionPolicy` specification [here](https://github.com/kubewarden/kubewarden-controller/blob/e0433fc3774d06dcf5e08bf2c600ad0117b89448/docs/crds/README.asciidoc#admissionpolicy).

Reviewer: ```suggestion

You can find the `AdmissionPolicy` specification [here](https://github.com/kubewarden/kubewarden-controller/blob/e0433fc3774d06dcf5e08bf2c600ad0117b89448/docs/crds/README.asciidoc#admissionpolicy).
```

---
title: Introducing the AdmissionPolicy
authors:
- Raul Cabello Martin
date: 2022-03-09
---

Up until now, the only way to define a policy in Kubewarden was to use a `ClusterAdmissionPolicy` which is applied to cluster-wide resources across all namespaces.

That's why we're thrilled to announce the new `AdmissionPolicy` resource. This new resource is created inside a `namespace` and the policies will process only the requests that are targeting the namespace where the `AdmissionPolicy` is defined. Except from being a "namespaced" resource, `AdmissionPolicy` works exactly the same as the `ClusterAdmissionPolicy`.

> ## Understanding the motivation for using AdmissionPolicies

Reviewer 3: Making the title more compelling.

```suggestion
## Why should you use AdmissionPolicies?
```

> It's possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector`, however there's no way for Kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources just in a particular namespace. 

Reviewer 3: ```suggestion
It's possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector`. However there's no way for Kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources just in a particular namespace. 
```

Allowing all tenants to deploy `ClusterAdmissionPolicies` is risky. A tenant could apply policies that affect resources in all namespaces, even if they don't have access to all of them.
You probably want to allow tenants to deploy policies just in the namespaces they have access to. That's where `AdmissionPolicies` come into play! We can achieve this restriction with `AdmissionPolicies` and Kubernetes’ Role-Based Access Control (RBAC).

> As an example, let's say you have a tenant who can deploy resources just to the `development` namespace. You can allow this tenant to deploy `AdmissionPolicies` just in this namespace using RBAC. If you allowed them to create `ClusterAdmissionPolicies`, they could block other resources from being handled in other namespaces.

Reviewer 3: Since we’re postulating that the tenant should be able to deploy only to the development namespace, slightly rephrasing the existing sentence.

```suggestion
As an example, let's say you want a tenant who can deploy resources just to the `development` namespace. You can allow this tenant to deploy `AdmissionPolicies` just in this namespace using RBAC. However, if you allowed them to create `ClusterAdmissionPolicies`, they could block other resources from being handled in other namespaces.
```

> ## Let's see an AdmissionPolicy in Action!

Reviewer 3: Shortening the title

```suggestion
## AdmissionPolicy in Action!
```

  namespace: production
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.1.9
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
  mutating: false

```

Wait for the policy to be active:

```
kubectl wait --for=condition=PolicyActive admissionpolicies ns-privileged-pods -n production
```

Create a file named `privileged-pod.yaml` with the following pod specification:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: privileged-pod
spec:
  containers:
    - name: nginx
      image: nginx:latest
      securityContext:
        privileged: true
```

Try to create a privileged pod in the `production` namespace:
```
kubectl apply -f privileged-pod.yaml -n production
```

You will get the following error:

```
Error from server: error when creating "privileged-pod.yaml": admission webhook "namespaced-kubewarden-privileged-pods.kubewarden.admission" denied the request: User 'my-user' cannot schedule privileged containers
```

Finally, verify you can successfully create a privileged pod in a different namespace:

```
kubectl apply -f privileged-pod.yaml -n development
```

## Conclusion

> As you could see, Kubewarden can now provide you with two different choices for deploying policies.

Reviewer 3: Slightly rephrasing the sentence for better readability

```suggestion
As evidenced in the above example, Kubewarden provides you with two different choices for deploying policies.
```

If your policy needs to be applied to resources across all namespaces or cluster-wide resources, then you can use `ClusterAdmissionPolicies`. 

> In the other hand, if your cluster is shared by multiple users or teams, using different namespaces or your policy needs to be applied only to resources within a namespace, then the new `AdmissionPolicy` will be the right choice.

Reviewer 3: Slight grammatical nit.

```suggestion
On the other hand, if your cluster is shared by multiple users or teams, uses different namespaces or your policy needs to be applied only to resources within a namespace, then the new `AdmissionPolicy` would be the right choice.
```

You can find the `AdmissionPolicy` specification [here](https://github.com/kubewarden/kubewarden-controller/blob/e0433fc3774d06dcf5e08bf2c600ad0117b89448/docs/crds/README.asciidoc#admissionpolicy).

> We would love to hear your feedback! Feel free to open issues in the

Reviewer 3: Grammatical nit

```suggestion
As a community, we strive on feedback and welcome your suggestions! Feel free to open an issue against our
```

> [github.com/kubewarden](https://github.com/kubewarden) projects or get in

Reviewer 3: ```suggestion
[GitHub repository](https://github.com/kubewarden/kubewarden-controller) or get in
```

> touch in the [#kubewarden Slack channel](https://kubernetes.slack.com/archives/C01T3GTC3L7)!

Reviewer 3: ```suggestion
touch on the [#kubewarden Slack channel](https://kubernetes.slack.com/archives/C01T3GTC3L7)!
```

---
title: Introducing the AdmissionPolicy
authors:
- Raul Cabello Martin
date: 2022-03-09
---

> Up until now, the only way to define a policy in Kubewarden was to use a `ClusterAdmissionPolicy`. `ClusterAdmissionPolicies` are applied to matching cluster-wide resources and resources across all namespaces.

Author: It doesn't apply anymore, see latest changes

We are thrilled to announce `AdmissionPolicy`, a namespaced resource. The policy will process only the requests that are targeting the namespace where the `AdmissionPolicy` is defined. Other than that, there are no functional differences between `AdmissionPolicy` and `ClusterAdmissionPolicy`.

## Understanding the motivation for using AdmissionPolicies

> It is possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector`, but there is no way for kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources just in a particular namespace. 

Reviewer 4: Nit (and I'm not sure if that's correct even)

```suggestion
It is possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector` (in K8s >= 1.21), but there is no way for Kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources in any namespace, once they are allowed to create them.
```

Author: `kubernetes.io/metadata.name` label, which is the namespace name, was introduced in k8s 1.21. I think you can still apply other filters with the namespaceSelector in versions below 1.21

Reviewer 4: Ook, resolving then.

> It is possible to restrict the namespaces where a `ClusterAdmissionPolicy` evaluates resources using a `namespaceSelector`, but there is no way for kubernetes administrators to restrict users from creating `ClusterAdmissionPolicies` that evaluate resources just in a particular namespace. 

Author: I would leave 'ClusterAdmissionPolicies' since a policy might be a `ClusterAdmissionPolicy` or an `AdmissionPolicy`

> Allowing all tenants to deploy `ClusterAdmissionPolicies` is risky. A tenant could apply policies that affect resources in all namespaces, even if he doesn't have access to all of them.

Reviewer 2: ```suggestion
Allowing all tenants to deploy `ClusterAdmissionPolicies` is risky. A tenant could apply policies that affect resources in all namespaces, even if they don't have access to all of them.
```

Author: done in https://github.com/kubewarden/kubewarden.io/pull/64/commits/8398829928bc5cc1b36718060fb70a2f84b024cd

> You probably want to allow tenants to deploy policies just in the namespaces they have access to. That's where `AdmissionPolicies` are useful! We can achieve this restriction with `AdmissionPolicies` and Kubernetes’s Role-Based Access Control (RBAC).

Reviewer 2: ```suggestion
You probably want to allow tenants to deploy policies just in the namespaces they have access to. That's where `AdmissionPolicies` are useful! We can achieve this restriction with `AdmissionPolicies` and Kubernetes' Role-Based Access Control (RBAC).
```

> Imagine we have a tenant who can deploy resources just to the development namespace. We can allow this tenant to deploy `AdmissionPolicies` just in the development namespace using RBAC. If we allowed him to create `ClusterAdmissionPolicies`, he could block other resources from being created in other namespaces.

Reviewer 4: Nit:

```suggestion
Imagine we have a tenant who can deploy resources just to the `development` namespace. We can allow this tenant to deploy `AdmissionPolicies` just in the `development` namespace using RBAC. If we allowed him to create `ClusterAdmissionPolicies`, he could block other resources from being created in other namespaces.
```

I would quote the namespace names in the rest of the post.

Author: done in https://github.com/kubewarden/kubewarden.io/pull/64/commits/8398829928bc5cc1b36718060fb70a2f84b024cd

> Imagine we have a tenant who can deploy resources just to the development namespace. We can allow this tenant to deploy `AdmissionPolicies` just in the development namespace using RBAC. If we allowed him to create `ClusterAdmissionPolicies`, he could block other resources from being created in other namespaces.

Reviewer 2: ```suggestion
Imagine we have a tenant who can deploy resources just to the `development` namespace. We can allow this tenant to deploy `AdmissionPolicies` just in this namespace using RBAC. If we allowed them to create `ClusterAdmissionPolicies`, they could block other resources from being handled in other namespaces.
```

Handled as policies can block any verb that Kubernetes allows.

## Let's see an AdmissionPolicy in Action!

> We are going to create an AdmissionPolicy that rejects privileged pods from being created in the production namespace.

Reviewer 2: ```suggestion
We are going to create an `AdmissionPolicy` that rejects privileged pods from being created in the `production` namespace.
```

> Assume we have a kubernetes cluster with Kubewarden installed. Follow the [quick start guide](https://docs.kubewarden.io/quick-start.html) if you don't have Kubewarden installed.

Reviewer 2: ```suggestion
Assume we have a Kubernetes cluster with Kubewarden installed. Follow the [quick start guide](https://docs.kubewarden.io/quick-start.html) if you don't have Kubewarden installed.
```

Author: done in https://github.com/kubewarden/kubewarden.io/pull/64/commits/8398829928bc5cc1b36718060fb70a2f84b024cd

> First, create two namespaces: development and production

Reviewer 4: Some of the lines have `:`, some not. I would add `:` to all for consistency.

Reviewer 2: Agree with @viccuad (made in a previous comment) that we should quote all namespaces names, here too, and in the rest of the occurences.

Author: done in https://github.com/kubewarden/kubewarden.io/pull/64/commits/8398829928bc5cc1b36718060fb70a2f84b024cd

```
kubectl create ns development 
kubectl create ns production
```

Let's create an AdmissionPolicy in the production namespace

```yaml
apiVersion: policies.kubewarden.io/v1alpha2
kind: AdmissionPolicy
metadata:
  name: ns-privileged-pods
  namespace: production
spec:
  module: registry://ghcr.io/kubewarden/policies/pod-privileged:v0.1.9
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
  mutating: false

```

Let's wait for the policy to be active

```
kubectl wait --for=condition=PolicyActive admissionpolicies ns-privileged-pods -n production
```

> Create a file named privileged-pod.yaml with the following pod specification:

Reviewer 2: ```suggestion
Create a file named `privileged-pod.yaml` with the following pod specification:
```

Author: done in https://github.com/kubewarden/kubewarden.io/pull/64/commits/8398829928bc5cc1b36718060fb70a2f84b024cd

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: privileged-pod
spec:
  containers:
    - name: nginx
      image: nginx:latest
      securityContext:
        privileged: true
```

If we try to create a privileged pod in the production namespace
```
kubectl apply -f privileged-pod.yaml -n production
```

we will get the following error

```
Error from server: error when creating "privileged-pod.yaml": admission webhook "namespaced-kubewarden-privileged-pods.kubewarden.admission" denied the request: User 'my-user' cannot schedule privileged containers
```

Finally, verify we can successfully create a privileged pod in a different namespace

```
kubectl apply -f privileged-pod.yaml -n development

> ```

Reviewer 2: I think it would be cool to perform a post closing paragraph by doing a summary of what we have seen and how `AdmissionPolicies` can help users in a summarized way.

Author: great idea! I added a conclusion [here](https://github.com/kubewarden/kubewarden.io/pull/64/commits/8398829928bc5cc1b36718060fb70a2f84b024cd) let me know what you think. Thanks!
